### DISCUSSION TITLE
Load cart using cart token

### DISCUSSION THREAD

**Discussion by gustavoconter** [woocommerce/woocommerce#44991] [WooCommerce, Blocks, and Site Editing]
I'm running a store with the CoCart plugin and thinking about switching to the Store API for more features. My main question is about loading a cart with a token, since I use the standard WooCommerce checkout and don't want to make a custom one.

From what I've read, it looks like I can either share cookies between my main site and the checkout site (like 'mystore.com' and 'checkout.mystore.com') or create a whole new checkout on my site. Can I load a cart right into the checkout? Is using shared cookies the best way to do this? I'd really appreciate any advice on this 🙏

**Comment by mikejolley** [comment]
Hi @gustavoconter 

The Cart-Token header doesn't require cookies. Responses from the API will return this token, and if you include it with requests it will return that same cart. 

> Can I load a cart right into the checkout?

I'm not 100% sure what you're asking here. If you have a custom checkout that runs by making requests to the Store API, then yes you can use the token instead.

**Reply by gustavoconter** [reply]
Hi @mikejolley, thanks for the reply.

Regarding the cart-token, it's working well. My main aim is to integrate the cart with WooCommerce's default checkout.

I'm planning to redirect users to my WooCommerce checkout and use a query or a rule to load the cart, like this:
`https://checkout.mystore.com/checkout?cart={{cart-token}}`.

In this case `https://mystore.com` is my headless store.

This URL would load the cart associated with a specific token.

I've created a code that allows the load of a specific cart using a query. This is just a prototype, and I'm not sure about potential security issues or other concerns. The concept is to intercept a token from the URL and load the corresponding cart.

Here's the code I used:

```php
<?php

use Automattic\WooCommerce\StoreApi\SessionHandler as StoreAPISessionHandler;
use Automattic\WooCommerce\StoreApi\Utilities\JsonWebToken as JWT;

/**
 * Register the cart query var so it can be used in the API.
 */
add_filter('query_vars', function ($vars) {
    $vars[] = "cart";
    return $vars;
});

$query_string = $_SERVER['QUERY_STRING'];
parse_str($query_string, $query_array);

$jwt_token = !empty($query_array['cart']) ? $query_array['cart'] : false;

if ($jwt_token) {
    $jwt_token_is_valid = JWT::validate($jwt_token, '@' . wp_salt());

    if ($jwt_token_is_valid) {
        $_SERVER['HTTP_CART_TOKEN'] = $jwt_token;

        add_filter(
            'woocommerce_session_handler',
            function ($handler) {
                return StoreAPISessionHandler::class;
            },
            10
        );
    }
}
```

**Reply by gustavoconter** [reply]
Also, for some strange reason if I run the filter `woocommerce_session_handler` inside a hook like `woocommerce_init` or even `init`, the entire code does not work, only if it runs directly.

The `query_vars` filter in this case is redundant.

**Reply by senadir** [reply]
The approach seems to be fine, I'd suggest that instead of loading `StoreAPISessionHandler`, you just figure out to which session the token belongs and assign it to the current user.

Otherwise your approach seems to be fine, I'm not sure why it's not working for `woocommerce_init` but it seems fine to run directly as long as you're using static methods for JWT.

**Reply by gustavoconter** [reply]
I will try to make it more robust and try to and understand why it does not work with `woocommerce_init`.
Thanks!

**Reply by gustavoconter** [reply]
Something that I can't understand why it happens is that If I try to load the cart on the checkout page, it gives me the "Your session has expired" notice.

```php
add_filter(
    'woocommerce_session_handler',
    function ($handler) {
        try {
            $query_string = $_SERVER['QUERY_STRING'];
            parse_str($query_string, $query_array);

            $jwt_token = !empty($query_array['cart']) ? $query_array['cart'] : false;

            if ($jwt_token) {
                $jwtTokenIsValid = JWT::validate($jwt_token, '@' . wp_salt());

                if ($jwtTokenIsValid) {
                    $_SERVER['HTTP_CART_TOKEN'] = $jwt_token;

                    $handler = StoreAPISessionHandler::class;

                    $GLOBALS['using_cart_token'] = true;
                }
            }
        } catch (Exception $e) {
        }

        return $handler;
    },
    100,
    2
);


// For some reason this code bypasses the "Session expired"
add_action('woocommerce_init', function () {
    if($GLOBALS['using_cart_token']) {
        
    }
});
```

If i add the `woocommerce_init` hook, even if it does not run any code. It removes the "Session expired" notice
